### PR TITLE
Improve the mangling of USRs

### PR DESCRIPTION
**Pull request by jrose-apple**
- Allow typealiases to use substitutions.
- Consistently use 'a' as a mangling operator.
- For generic typealiases, include the alias as context for any generic parameters.

Typealiases don't show up in symbol names, which always refer to canonical types, but they are mangled for debug info and for USRs (unique identifiers used by SourceKit), so it's good to get this right.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/AST/ASTMangler.cpp
Comment (slavapestov) [inline lib/AST/ASTMangler.cpp]: Maybe we should add a getUnboundGenericType() to TypeDecl. You already had the same bit of code in an earlier PR today, and I've written it out in a few places also.

File: lib/AST/ASTMangler.cpp
Comment (slavapestov) [inline lib/AST/ASTMangler.cpp]: Is this an accident?

File: lib/Demangling/Demangler.cpp
Comment (slavapestov) [inline lib/Demangling/Demangler.cpp]: Maybe rename demangleNominalType()?

File: lib/AST/ASTMangler.cpp
Comment (jrose-apple) [inline lib/AST/ASTMangler.cpp]: Fine with me. Does it return UnboundGenericType, or just Type to handle the non-generic case?

File: lib/AST/ASTMangler.cpp
Comment (jrose-apple) [inline lib/AST/ASTMangler.cpp]: Oops, yes, that's the next patch I'm working on. Thanks.

File: lib/Demangling/Demangler.cpp
Comment (jrose-apple) [inline lib/Demangling/Demangler.cpp]: I wish we had a better name for "nominal or typealias" than "generic type".

**Comment by jrose-apple** [comment]
About to push fixes for the simple things, but I figure I can let the tests keep going while I'm waiting for more feedback: [Linux](https://ci.swift.org/job/swift-PR-Linux/6791/), [Mac](https://ci.swift.org/job/swift-PR-osx/6814/)

File: lib/AST/ASTMangler.cpp
Comment (slavapestov) [inline lib/AST/ASTMangler.cpp]: I think the latter. No need to do that before merging the PR of course.

**Comment by jrose-apple** [comment]
Both full tests passed, and the only things since then have been renames and removing commented-out code.

@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Added additional verification, fixed a bug that would have been caught by that verification, and updated ABI.rst.

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@shahmishal, any idea why the failure isn't getting posted back on GitHub?

**Comment by shahmishal** [comment]
It should have been 82e76edbb7170b0fce95eed922fb7e50bfea9bc0 not c7f9cb7548c51da7db2d21efe7d28478c8ee8e78.
```
Setting status of c7f9cb7548c51da7db2d21efe7d28478c8ee8e78 to FAILURE with url https://ci.swift.org/job/swift-PR-Linux/6815/ and message: '**Build failed**
```

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
:-( The builder is testing the correct commit but GitHub is showing them out of order.

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

**Comment by jrose-apple** [comment]
Previous full-test passed. Still waiting on review from a SourceKit person. (Erik and I have been talking through chat.)

@swift-ci Please smoke test

**Comment by akyrtzi** [comment]
Changes LGTM.

**Review by eeckstein — APPROVED** [review]
LGTM!

**Pull request merged**
